### ISSUE TITLE
About test accuracy during experiment

### ISSUE THREAD

**Issue by DiahannWu** [dayu11/Differentially-Private-Deep-Learning#4]
cifar_train.py  --eps 8 --delta 1e-5 --rank 16 --lr 0.5 --clipping 1 --batchsize 1000 --n_epoch 400 --width 10
Using the above parameters to train the cifar10 dataset under RGP, the accuracy is just over 40%. How to deal with it?
Test environment: win10, cuda11.1
I run it directly in pycharm IDE with the following settings:

    parser.add_argument('--dataset', default='cifar10', type=str, help='dataset name')
    parser.add_argument('--arch', default='resnet28', type=str, help='model name')  # resnet28
    parser.add_argument('--resume', '-r', action='store_true', help='resume from checkpoint')
    parser.add_argument('--sess', default='resnet', type=str, help='session name')
    parser.add_argument('--weight_decay', default=1e-4, type=float, help='weight decay (default=1e-4)')
    parser.add_argument('--batchsize', default=1000, type=int, help='batch size')
    parser.add_argument('--n_epoch', default=400, type=int, help='total number of epochs')
    parser.add_argument('--lr', default=0.5, type=float, help='base learning rate (default=0.4)')
    parser.add_argument('--momentum', default=0.9, type=float, help='momentum coeeficient')

    parser.add_argument('--eps', default=8, type=float, help='eps value')
    parser.add_argument('--width', default=10, type=int, help='model width')
    parser.add_argument('--delta', default=1e-5, type=float, help='delta value')
    parser.add_argument('--rank', default=16, type=int, help='rank of reparameterization')
    parser.add_argument('--clipping', default=1., type=float, help='clipping threshold')
    parser.add_argument('--warmup_epoch', default=-1, type=int, help='num. of epochs for warmup')

**Comment by dayu11** [comment]
can you paste your training log here? I will take a look. I will also re-clone the repo and run this command for a check
```
CUDA_VISIBLE_DEVICES=0 python cifar_train.py  --eps 8 --delta 1e-5 --rank 16 --lr 0.5 --clipping 1 --batchsize 1000 --n_epoch 400 --width 10
```

**Comment by DiahannWu** [comment]
您发送的邮件我已经收到，谢谢

**Comment by dayu11** [comment]
```
Epoch: 390
Train loss:1.05560 time: 60 s train acc: 71.058 test loss:1.18920 test acc: 68.33

Epoch: 391
Train loss:1.05669 time: 60 s train acc: 71.016 test loss:1.19185 test acc: 68.49

Epoch: 392
Train loss:1.05714 time: 60 s train acc: 71.054 test loss:1.18815 test acc: 68.48

Epoch: 393
Train loss:1.05554 time: 60 s train acc: 71.09 test loss:1.19196 test acc: 68.41

Epoch: 394
Train loss:1.05650 time: 60 s train acc: 71.088 test loss:1.18953 test acc: 68.43

Epoch: 395
Train loss:1.05511 time: 60 s train acc: 71.114 test loss:1.18947 test acc: 68.49

Epoch: 396
Train loss:1.05617 time: 60 s train acc: 71.12 test loss:1.18817 test acc: 68.55

Epoch: 397
Train loss:1.05558 time: 60 s train acc: 71.096 test loss:1.18956 test acc: 68.42

Epoch: 398
Train loss:1.05545 time: 60 s train acc: 71.172 test loss:1.18805 test acc: 68.33

Epoch: 399
Train loss:1.05430 time: 60 s train acc: 71.1 test loss:1.19154 test acc: 68.41
```

Here is the final accuracy I get. The run takes ~5 hours on a single A100 GPU.

**Comment by DiahannWu** [comment]
this is the accuracy on my device.  It takes 10+ hours on GeForce RTX 3090 but the runs is not over yet
I wonder whether the calculation will vary greatly in different systems?

noise standard deviation for eps = 8.0:  2.0840000000000014
number of parameters (low-rank): 0.152 M number of parameters (full): 36.414 M

```
Epoch: 0
Train loss:2.21895 time: 192 s train acc: 17.316 test loss:2.05727 test acc: 24.3

Epoch: 1
Train loss:1.94833 time: 220 s train acc: 29.094 test loss:1.84615 test acc: 34.08

Epoch: 2
Train loss:1.80164 time: 231 s train acc: 35.092 test loss:1.73646 test acc: 38.2

Epoch: 3
Train loss:1.72130 time: 230 s train acc: 38.268 test loss:1.68401 test acc: 40.13

Epoch: 4
Train loss:1.68326 time: 232 s train acc: 40.154 test loss:1.65855 test acc: 41.23

Epoch: 5
Train loss:1.66585 time: 232 s train acc: 41.438 test loss:1.65947 test acc: 41.96

Epoch: 6
Train loss:1.66452 time: 231 s train acc: 42.068 test loss:1.67196 test acc: 41.93

Epoch: 7
Train loss:1.67372 time: 232 s train acc: 42.102 test loss:1.68782 test acc: 42.14

Epoch: 8
Train loss:1.68670 time: 220 s train acc: 42.24 test loss:1.70304 test acc: 42.38

Epoch: 9
Train loss:1.70462 time: 214 s train acc: 42.298 test loss:1.72826 test acc: 42.14

Epoch: 10
Train loss:1.72768 time: 215 s train acc: 42.22 test loss:1.74814 test acc: 42.17

Epoch: 178
Train loss:1.86448 time: 224 s train acc: 43.902 test loss:1.93878 test acc: 41.97

Epoch: 179
Train loss:1.86471 time: 224 s train acc: 43.834 test loss:1.93942 test acc: 41.85

Epoch: 180
Train loss:1.86465 time: 222 s train acc: 43.872 test loss:1.93892 test acc: 41.94

Epoch: 181
Train loss:1.86444 time: 225 s train acc: 43.87 test loss:1.93712 test acc: 41.87

Epoch: 182
Train loss:1.86429 time: 219 s train acc: 43.86 test loss:1.93809 test acc: 41.93

Epoch: 183
Train loss:1.86431 time: 219 s train acc: 43.912 test loss:1.93973 test acc: 41.9

Epoch: 184
Train loss:1.86454 time: 219 s train acc: 43.906 test loss:1.93856 test acc: 42.01

Epoch: 185
Train loss:1.86427 time: 218 s train acc: 43.882 test loss:1.93896 test acc: 41.97

Epoch: 186
Train loss:1.86445 time: 223 s train acc: 43.852 test loss:1.93728 test acc: 41.99

Epoch: 187
Train loss:1.86442 time: 232 s train acc: 43.896 test loss:1.93951 test acc: 41.98
```

**Comment by dayu11** [comment]
“I wonder whether the calculation will vary greatly in different systems?”

It should not be. It's wired that the loss at epoch 10 is even smaller than that at epoch 180. Did you adjust the learning rate properly? See Line 334 at RGP/cifar_train.py

**Comment by DiahannWu** [comment]
I adjust the learning rate just the same with you, as follows:

    def adjust_learning_rate(optimizer, epoch):
        if(epoch<0.3*args.n_epoch):
            decay = 1.
        elif(epoch<0.6*args.n_epoch):
            decay = 5.
        elif(epoch<0.8*args.n_epoch):
            decay = 25.
        else:
            decay = 125.

when initial lr=0.5, training runs=400, Is the following noise standard deviation the same as your experiment? 
''noise standard deviation for eps = 8.0: 2.0840000000000014''

When I set the epochs=200, the calculation of  noise standard deviation will be smaller, and test acc will be improved, but still pool, result as follows:

noise standard deviation for eps = 8.0:  1.5570000000000008

```
epoch,lr,train loss,train acc,test loss,test acc

0,0.5,2.2866114304990184,17.286,2.1243897749453176,25.43

1,0.5,1.9933857236589705,29.348,1.878606662458303,33.79

2,0.5,1.8278235902591629,35.824,1.7372706568970973,39.12

3,0.5,1.7250678466290843,39.61,1.6821480381245515,41.78

4,0.5,1.671153494289943,42.048,1.6501511675970895,43.47

5,0.5,1.6455494287062664,43.22,1.637024069318966,44.06

6,0.5,1.637631216827704,43.894,1.6353588396189165,44.58

7,0.5,1.6405540704727173,44.184,1.6401693601997531,44.98

8,0.5,1.647700959322404,44.622,1.6416832889829363,45.14

9,0.5,1.6572064015330101,44.856,1.6469415499239553,45.41

10,0.5,1.6656412956665974,45.2,1.6621089322226388,45.53

115,0.1,1.7489330063060837,47.382,1.824404322371191,45.19

116,0.1,1.7487009082521712,47.388,1.824752420795207,45.15

117,0.1,1.7485986315474218,47.386,1.8256115475479437,45.11

118,0.1,1.7484297606409813,47.358,1.825326141046018,45.16

119,0.1,1.748225820307829,47.348,1.82511960973545,45.24

120,0.02,1.745273341937941,47.418,1.8245798762963743,45.3

121,0.02,1.7448297447087813,47.458,1.82479448464452,45.25

122,0.02,1.744821726059427,47.426,1.8245343894374615,45.3
```
Then, I'll try to use weight decay in earlier epochs

**Comment by dayu11** [comment]
yeah the noise std is the same as mine. the code uses a weight_decay=1e-4 by default

**Comment by DiahannWu** [comment]
I just said wrong, I also use weight_decay=1e-4, actually, I use all the hyperparameters the same with you
I just mean I'll try to use lr decay in earlier epochs, and I don't know whether it will work, and my training is very time-consuming
I saw a link(https://zhuanlan.zhihu.com/p/242431161) about different types of data between different environment, will this affect the result? does it means in different types of GPU, the types of data in GPU's calculation will be different

**Comment by dayu11** [comment]
I personally do not think the types of GPU will have such a big influence. I have run this code on P100, V100, and A100 and all the results are similar.

**Comment by dayu11** [comment]
You can push your code to a public github repo and share the link here. I will take a look and try to help

**Comment by dayu11** [comment]
> I just said wrong, I also use weight_decay=1e-4, actually, I use all the hyperparameters the same with you I just mean I'll try to use lr decay in earlier epochs, and I don't know whether it will work, and my training is very time-consuming I saw a link(https://zhuanlan.zhihu.com/p/242431161) about different types of data between different environment, will this affect the result? does it means in different types of GPU, the types of data in GPU's calculation will be different

Hmm but I have not run this code on Windows. There is a slight chance that this difference causes the problem. Maybe try a different random seed and larger weight_decay (say 5e-4 or 1e-3)? As I can recall, using a larger weight_decay further improves the performance.

**Comment by DiahannWu** [comment]
> 

[this is the code in my environment](https://github.com/DiahannWu/RGP_ofYU), the code has hardly changed, now I'll try the weight_decay you suggest

**Comment by dayu11** [comment]
> > 
> 
> [this is the code in my environment](https://github.com/DiahannWu/RGP_ofYU), the code has hardly changed, now I'll try the weight_decay you suggest

Ok things are getting weird. I ran your code on my hardware and get outputs as expected


```
Epoch: 390
Train loss:1.06750 time: 60 s train acc: 70.568 epoch 390 0.004
test loss:1.19690 test acc: 68.22

Epoch: 391
Train loss:1.06672 time: 60 s train acc: 70.538 epoch 391 0.004
test loss:1.19644 test acc: 68.24

Epoch: 392
Train loss:1.06641 time: 60 s train acc: 70.536 epoch 392 0.004
test loss:1.19607 test acc: 68.27

Epoch: 393
Train loss:1.06706 time: 60 s train acc: 70.544 epoch 393 0.004
test loss:1.19363 test acc: 68.42

Epoch: 394
Train loss:1.06615 time: 60 s train acc: 70.576 epoch 394 0.004
test loss:1.19730 test acc: 68.18

Epoch: 395
Train loss:1.06716 time: 60 s train acc: 70.548 epoch 395 0.004
test loss:1.19462 test acc: 68.32

Epoch: 396
Train loss:1.06639 time: 60 s train acc: 70.508 epoch 396 0.004
test loss:1.19568 test acc: 68.29

Epoch: 397
Train loss:1.06699 time: 60 s train acc: 70.532 epoch 397 0.004
test loss:1.19562 test acc: 68.28

Epoch: 398
Train loss:1.06570 time: 60 s train acc: 70.562 epoch 398 0.004
test loss:1.19705 test acc: 68.25

Epoch: 399
Train loss:1.06564 time: 60 s train acc: 70.602 epoch 399 0.004
test loss:1.19667 test acc: 68.23
```


Two minor things I modified are changing 'cuda:1' to 'cuda' and print learning rate at every epoch. Maybe this is really a hardware issue. I will try to run this code on windows at some point

**Comment by DiahannWu** [comment]
When using weight_decay=1e-3, test accuracy is up to ~52
```
Epoch: 119
Train loss:1.46729 time: 236 s train acc: 51.456 test loss:1.47489 test acc: 50.87

Epoch: 120
Train loss:1.40587 time: 236 s train acc: 52.77 test loss:1.43018 test acc: 51.91

Epoch: 121
Train loss:1.39460 time: 232 s train acc: 53.224 test loss:1.43809 test acc: 51.98

Epoch: 122
Train loss:1.39642 time: 234 s train acc: 53.256 test loss:1.43777 test acc: 52.04

Epoch: 123
Train loss:1.39704 time: 235 s train acc: 53.302 test loss:1.43597 test acc: 52.03

Epoch: 124
Train loss:1.39740 time: 235 s train acc: 53.428 test loss:1.43569 test acc: 51.99

Epoch: 125
Train loss:1.39788 time: 235 s train acc: 53.45 test loss:1.43513 test acc: 52.0

Epoch: 126
Train loss:1.39835 time: 234 s train acc: 53.482 test loss:1.43435 test acc: 51.99

Epoch: 127
Train loss:1.39866 time: 235 s train acc: 53.444 test loss:1.43414 test acc: 52.0

Epoch: 128
Train loss:1.39940 time: 232 s train acc: 53.43 test loss:1.43487 test acc: 51.98

Epoch: 129
Train loss:1.39973 time: 233 s train acc: 53.408 test loss:1.43571 test acc: 51.97

Epoch: 130
Train loss:1.40028 time: 235 s train acc: 53.352 test loss:1.43595 test acc: 51.92

Epoch: 131
Train loss:1.40064 time: 233 s train acc: 53.352 test loss:1.43634 test acc: 52.01

Epoch: 132
Train loss:1.40118 time: 234 s train acc: 53.31 test loss:1.43671 test acc: 52.06

Epoch: 133
Train loss:1.40178 time: 235 s train acc: 53.286 test loss:1.43731 test acc: 52.08

Epoch: 134
Train loss:1.40229 time: 232 s train acc: 53.326 test loss:1.43774 test acc: 52.07

Epoch: 135
Train loss:1.40271 time: 232 s train acc: 53.29 test loss:1.43819 test acc: 52.08

Epoch: 136
Train loss:1.40326 time: 234 s train acc: 53.278 test loss:1.43859 test acc: 52.02

Epoch: 137
Train loss:1.40391 time: 235 s train acc: 53.246 test loss:1.43937 test acc: 52.06

Epoch: 138
Train loss:1.40446 time: 234 s train acc: 53.194 test loss:1.44054 test acc: 52.13

Epoch: 139
Train loss:1.40490 time: 234 s train acc: 53.2 test loss:1.44095 test acc: 52.09

Epoch: 140
Train loss:1.40534 time: 234 s train acc: 53.194 test loss:1.44272 test acc: 51.93

Epoch: 141
Train loss:1.40605 time: 232 s train acc: 53.162 test loss:1.44379 test acc: 51.9
```

**Comment by dayu11** [comment]
> When using weight_decay=1e-3, test accuracy is up to ~52
> 
> ```
> Epoch: 119
> Train loss:1.46729 time: 236 s train acc: 51.456 test loss:1.47489 test acc: 50.87
> 
> Epoch: 120
> Train loss:1.40587 time: 236 s train acc: 52.77 test loss:1.43018 test acc: 51.91
> 
> Epoch: 121
> Train loss:1.39460 time: 232 s train acc: 53.224 test loss:1.43809 test acc: 51.98
> 
> Epoch: 122
> Train loss:1.39642 time: 234 s train acc: 53.256 test loss:1.43777 test acc: 52.04
> 
> Epoch: 123
> Train loss:1.39704 time: 235 s train acc: 53.302 test loss:1.43597 test acc: 52.03
> 
> Epoch: 124
> Train loss:1.39740 time: 235 s train acc: 53.428 test loss:1.43569 test acc: 51.99
> 
> Epoch: 125
> Train loss:1.39788 time: 235 s train acc: 53.45 test loss:1.43513 test acc: 52.0
> 
> Epoch: 126
> Train loss:1.39835 time: 234 s train acc: 53.482 test loss:1.43435 test acc: 51.99
> 
> Epoch: 127
> Train loss:1.39866 time: 235 s train acc: 53.444 test loss:1.43414 test acc: 52.0
> 
> Epoch: 128
> Train loss:1.39940 time: 232 s train acc: 53.43 test loss:1.43487 test acc: 51.98
> 
> Epoch: 129
> Train loss:1.39973 time: 233 s train acc: 53.408 test loss:1.43571 test acc: 51.97
> 
> Epoch: 130
> Train loss:1.40028 time: 235 s train acc: 53.352 test loss:1.43595 test acc: 51.92
> 
> Epoch: 131
> Train loss:1.40064 time: 233 s train acc: 53.352 test loss:1.43634 test acc: 52.01
> 
> Epoch: 132
> Train loss:1.40118 time: 234 s train acc: 53.31 test loss:1.43671 test acc: 52.06
> 
> Epoch: 133
> Train loss:1.40178 time: 235 s train acc: 53.286 test loss:1.43731 test acc: 52.08
> 
> Epoch: 134
> Train loss:1.40229 time: 232 s train acc: 53.326 test loss:1.43774 test acc: 52.07
> 
> Epoch: 135
> Train loss:1.40271 time: 232 s train acc: 53.29 test loss:1.43819 test acc: 52.08
> 
> Epoch: 136
> Train loss:1.40326 time: 234 s train acc: 53.278 test loss:1.43859 test acc: 52.02
> 
> Epoch: 137
> Train loss:1.40391 time: 235 s train acc: 53.246 test loss:1.43937 test acc: 52.06
> 
> Epoch: 138
> Train loss:1.40446 time: 234 s train acc: 53.194 test loss:1.44054 test acc: 52.13
> 
> Epoch: 139
> Train loss:1.40490 time: 234 s train acc: 53.2 test loss:1.44095 test acc: 52.09
> 
> Epoch: 140
> Train loss:1.40534 time: 234 s train acc: 53.194 test loss:1.44272 test acc: 51.93
> 
> Epoch: 141
> Train loss:1.40605 time: 232 s train acc: 53.162 test loss:1.44379 test acc: 51.9
> ```

I think weight_decay is not the main problem here, as I can reproduce the results on my hardware with your code.

**Comment by DiahannWu** [comment]
您发送的邮件我已经收到，谢谢

**Comment by dayu11** [comment]
Please let me know if you still need help.

**Comment by DiahannWu** [comment]
您发送的邮件我已经收到，谢谢
